test/test_c10d.py

```diff
@@ -2831,6 +2848,83 @@ def _test_broadcast_coalesced(self, process_group, device):
 
         self.assertEqual(tensors, target)
 
+    def _run_all_reduce(self, pg):
+        pg.allreduce(torch.rand(10).cuda(self.rank))
+
+    @requires_nccl()
+    @skip_if_not_multigpu
+    def test_nccl_errors_nonblocking(self):
+        store = c10d.FileStore(self.file.name, self.world_size)
+        process_group = c10d.ProcessGroupNCCL(store, self.rank, self.world_size)
+        process_group.allreduce(torch.rand(10).cuda(self.rank))
+        if self.rank == 0:
+            # Should run fine without raising any errors.
+            work = process_group.allreduce(torch.rand(10).cuda(self.rank))
+            work.wait()
+
+            # Now the work scheduled next should hang forever since the previous
```

Comment:
Curious, which line does it hang at? Is it exactly [here](https://github.com/jeffdaily/pytorch/blob/f43b725472406ec1cfb35fe82dd3cc93ed05fcf7/torch/lib/c10d/ProcessGroupNCCL.cpp#L434) when calling `ncclAllReduce`? If so, it seems it does not only enqueue the op, but also perform some checking on the `NcclComm`.